crates/compiler/types/src/types.rs

```diff
@@ -333,6 +339,40 @@ impl OptAbleType {
     }
 }
 
+/// Polarity of a type, or roughly, what side of an arrow it appears on.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum Polarity {
+    /// A type that appears in negative/input position
+    Neg,
+    /// A type that appears in positive/output position
+    Pos,
+}
+
+impl std::ops::Neg for Polarity {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        match self {
+            Polarity::Neg => todo!(),
+            Polarity::Pos => todo!(),
+        }
+    }
+}
+
+impl Polarity {
+    pub const OF_VALUE: Polarity = Polarity::Pos;
+
+    pub const OF_PATTERN: Polarity = Polarity::Neg;
```

Comment:
Because our polarity is determined on the side of an arrow you're on this is faithful